### PR TITLE
rewrite cephdisks / hardware detection ref #70

### PR DESCRIPTION
**Pull request by jschmid1**
[WIP]
Signed-off-by: Joshua Schmid <jschmid@suse.de>

### CODE REVIEW

**Comment by jschmid1** [comment]
test this please

**Comment by jschmid1** [comment]
retest this please

**Comment by jschmid1** [comment]
retest this please

**Comment by jschmid1** [comment]
retest this please

**Comment by jschmid1** [comment]
test this please

**Comment by jschmid1** [comment]
test this please

**Comment by jschmid1** [comment]
test this please

File: srv/salt/_modules/cephdisks.py [hunk 1]
```diff
@@ -2,82 +2,429 @@
...
+        kwargs adds the possibility to pass custom raidcontroller
+        names to the class.
+        args:
+            hw_raid(bool): Manually set hw_raid True if this class can't detect it
+            hw_raid_name(str): Manually set the hw_raid_ctrls name
+            software_raid(bool): Manually set if you have sw raid and the class
+                                            fails to detect it.
+        """
+        self.overwrite_detection_method = kwargs.get('detection_method', None)
+        self.detection_method = self.find_detection_tool()
```
Comment (jan--f) [inline srv/salt/_modules/cephdisks.py]: Do you have a reason to add the ```overwrite_detection_method``` field to the object? It is only used by ```find_detection_tool```, so passing it as a method parameter seems more suitable to me.

    self.detection_method = self.find_detection_tool(kwargs.get('detection_method', None))

File: srv/salt/_modules/cephdisks.py [hunk 2]
```diff
@@ -2,82 +2,429 @@
...
+                exe_file = os.path.join(path, program)
+                if is_exe(exe_file):
+                    return exe_file
+        if nofail is True:
+            return None
+        elif nofail is False:
+            raise StandardError("Can't find the tool: {}. Please Install it in order to resume.".format(program))
+        else:
+            raise StandardError("nofail needs to be bool(True) or bool(False)")
+
```
Comment (jan--f) [inline srv/salt/_modules/cephdisks.py]: There's a bunch of unnecessary whitespaces here. I'm always a fan of removing these.

File: srv/salt/_modules/cephdisks.py (hunk 1 above)
Comment (jschmid1) [inline srv/salt/_modules/cephdisks.py]: good catch, will update that.

File: srv/salt/_modules/cephdisks.py (hunk 2 above)
Comment (jschmid1) [inline srv/salt/_modules/cephdisks.py]: done

**Comment by jan--f** [comment]
@jschmid1 can you please rebase onto master. Currently your branch fails in stage 1 with an exeption: ```NameError: global name '_CephCluster__create_dirs' is not defined```

**Comment by jschmid1** [comment]
done @jan--f

**Comment by jan--f** [comment]
ok get to stage 3 after now. Don't get a healthy cluster though...stage 3 is stuck for the past ~2h, MONs are up, no OSDs.
```
salt-run jobs.active
20170308115416927433:
    ----------
    Arguments:
        - ceph.pool
        |_
          ----------
          __kwarg__:
              True
          concurrent:
              False
          queue:
              False
          saltenv:
              base
    Function:
        state.sls
    Returned:
    Running:
        |_
          ----------
          admin:
              16308
    StartTime:
        2017, Mar 08 11:54:16.927433
    Target:
        admin
    Target-type:
        glob
    User:
        salt
```

On a purged cluster...will destroy and recreate.

**Comment by swiftgist** [comment]
I initially ran into an issue since I was overriding packages/common where a missing lspci fails quietly.  I am debating what the default behavior should be.  Part of me would like the message to be more prominent, but I know it's a separate module.  The other option could be to return what was found.  (I don't know why a Linux distribution would not have lspci (maybe repo trouble), but some list would be better than no list.)

I will say that cephdisks.list_drives bugs me. :)  I think the issue you wanted to avoid was overriding keywords.  I should have done what's listed at https://docs.saltstack.com/en/latest/ref/modules/#aliasing-functions.

**Comment by jan--f** [comment]
Ok works for me. I'd still like to see most methods in cephdisks.py as private though. Unless there is a good reason not to?

**Comment by jschmid1** [comment]
>  The other option could be to return what was found. (I don't know why a Linux distribution would not have lspci (maybe repo trouble), but some list would be better than no list.)

The vagrant box does not come with lspci/lsscsi preinstalled, so I had to add it to packages.common.
Usually these packages should be present.

> I will say that cephdisks.list_drives bugs me. :) I think the issue you wanted to avoid was overriding keywords. I should have done what's listed at https://docs.saltstack.com/en/latest/ref/modules/#aliasing-functions.

Correct, I wanted to avoid overwriting _list_ 

-> looks good!
```
__func_alias__ = {
    'set_': 'set',
    'list_': 'list',
}
```

I'll adapt to that in the next commit.

**Comment by jschmid1** [comment]
> I initially ran into an issue since I was overriding packages/common where a missing lspci fails quietly.

Oh, true. Haven't thought about the fact that common.packages is pointing to default-zypper.sls rather than default.sls hence made my changes in the latter..

Will fix that.

We still should track the behavior of pkg.install. It should support zyppers  "--no-gpg-checks" eventually to make our lives easier.

**Pull request merged**